Feature/outer sdk

BarutJozef opened the pull request:
This PR is an updated and full version of the OuterRail SDK

Notable changes:
- SDK - Library code, documentation, publishing related files
- Backend - REMOVING the API_KEY auth option + fixes for bugs found while testing SDK
- Frontend - Minor updates concerning Docs for the SDK

I request a review mostly on the the BACKEND updates, if they are OK, or I should update / change anything else.

marcel-vesely-kinit reviewed (commented):
I have reviewed the `backend/` and `frontend/` folders and intentionally skipped the `sdk-py/` folder due to obvious reasons. I've also pushed some changes to appease linter's pleas and pytest checks.

I still have yet to check the functionality of SDK as a whole, but I gave you some things to work on.

One more thing, since we're no longer using API keys, we should get rid of all the logic pertaining to them (backend code, frontend code, ...)
- Notice: Don't change the RailUser database model, keep adhering to its previous schema with the API key, but it will simply not be used anymore.

marcel-vesely-kinit commented:
Other things to change:
- Change the minimum Python version required for the OuterRail SDK to run: In documentation you state Python 3.9 should suffice, yet I wasn't able to spin it up even with 3.10, 3.11 was necessary
- In some snippets and within the OuterRail SDK documentation you use this particular way of configuring AIoD connection: `config = Configuration(host="https://rail.aiod.eu/api/docs")` -> this is incorrect due to multiple reasons
  - You need to remove the `/docs` suffix from the URL to actually point to the RAIL API
  - Default value of `auth_client_id` argument used by the `Configuration` is `outer-rail-sdk` that is outdated

marcel-vesely-kinit commented on sdk-py/OuterRail/instances/experiment_run.py:
```diff
@@ -0,0 +1,199 @@
+            except Exception as e:
+                raise e
+
+    def download_file(self, filepath: str, to_dir: str) -> None:
```
This function doesn't seem to be working properly. I tried downloading both textual and binary files and there seems to be some problems with both of them.

When downloading a textual file, this code crashes at the line 116 (`f.write(data)`) as the function expects a string, even though the data is a binary. When I tried changing the mode from `w` to `wb`, the problem was partialy solved: the code did work to completion, but the output text file was a binary gibberish.

When downloading a binary file (`.pkl` for instance), this code crashes even sooner with the following error: `UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte`

marcel-vesely-kinit commented:
One more thing: We can comment out the create dataset functionality if you want to. We don't support it on frontend (due to some reasons I've already forgotten), so unless it runs without a hitch, we can simply comment it out for the time being.

marcel-vesely-kinit commented on frontend/src/app/docs/docs.component.html:
```diff
@@ -11,7 +11,7 @@
                 <mat-divider></mat-divider>
 
                 <a mat-list-item routerLink="rail-sdks" routerLinkActive="active">RAIL SDKs</a>
```
I'd add analogical non-clickable navigation item called "Intro to RAIL" for the RAIL pages as well (What is RAIL?, Main concepts, ...)

BarutJozef commented:
After recent commits, all found issues and suggestions should be fixed. Token persistence should be less error prone as well, although I found no issue while testing.

marcel-vesely-kinit commented on sdk-py/OuterRail/instances/dataset.py:
```diff
@@ -0,0 +1,352 @@
+    application_area: Optional[List[StrictStr]] = Field(
+        default=None, description="The objective of this AI resource."
+    )
+    citation: Optional[List[StrictInt]] = Field(
```
Update all the AIoD asset data models used in the SDK to correspond to their latest version that can also be found in the `app/schemas/aiod_generated.py` file.

In this particular example, the `citation` attribute lists AIoD IDs of publications citing this asset. So instead of being `StrictInt` it should be `StrictStr`

marcel-vesely-kinit replied:
Don't forget to check the validity of data types of all the attributes

marcel-vesely-kinit reviewed (changes requested):
THINGS TO DO ONCE EVERYTHING IS RESOLVED:
- Update Github pages
- Release the newest SDK version to PyPI

BarutJozef commented:
> There are two mistakes in this change:
> 
> * You still use `auth_host` variable rather than `self.auth_host` attribute of the class within the `KeycloackOpenID` constructor
> * The condition is flawed: If the `auth_host` ends with "/", the self.auth_host is assigned an empty string

Fixed the ternary operator and Keycloak constructor now recieves class attributes.


> Doesn't seem to work.
> 
> After having switched from one Keycloak version to another, I got an error stating the `Configuration` class doesn't have a `token` attribute (error on this particular line).

self.token was actually not present in here and the token loaded from file not yet assigned to the class should have been invalidated instead. It is now fixed.


> Update all the AIoD asset data models used in the SDK to correspond to their latest version that can also be found in the `app/schemas/aiod_generated.py` file.
> 
> In this particular example, the `citation` attribute lists AIoD IDs of publications citing this asset. So instead of being `StrictInt` it should be `StrictStr`

I crossed check all models present in the SDK and updated them to match Backend schemas.
